app/uk/gov/hmrc/cgtpropertydisposalsfrontend/controllers/actions/AuthenticatedAction.scala

```diff
@@ -71,48 +78,82 @@ class AuthenticatedAction @Inject() (
       (selfBaseUrl + successRelativeUrl) -> (selfBaseUrl + failureRelativeUrl)
   }
 
-  @SuppressWarnings(Array("org.wartremover.warts.Product", "org.wartremover.warts.Serializable"))
-  override protected def refine[A](request: MessagesRequest[A]): Future[Either[Result, AuthenticatedRequest[A]]] = {
-    implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromHeadersAndSession(request.headers, Some(request.session))
-
-    authorisedFunctions.authorised(ConfidenceLevel.L200).retrieve(Retrievals.nino) {
-      case None =>
-        logger.warn("Could not find NINO")
+  private def makeAuthenticatedRequestWithItmpRetrieval[A](
+    nino: String,
+    name: ItmpName,
+    dateOfBirth: LocalDate,
+    request: MessagesRequest[A]
+  ): Future[Either[Result, AuthenticatedRequest[A]]] =
+    name match {
+      case ItmpName(Some(givenName), _, Some(familyName)) =>
+        Right(AuthenticatedRequest(NINO(nino), Name(givenName, familyName), DateOfBirth(dateOfBirth), request))
+      case other =>
+        logger.warn(s"Failed to retrieve the complete name for this user: $other")
         Left(errorHandler.errorResult()(request))
+    }
 
-      case Some(nino) =>
-        Right(AuthenticatedRequest(NINO(nino), request))
-    }.recoverWith {
-      case _: NoActiveSession =>
-        Left(Redirect(signInUrl, Map("continue" -> Seq(selfBaseUrl + request.uri), "origin" -> Seq(origin))))
-
-      case _: InsufficientConfidenceLevel =>
-        handleInsufficientConfidenceLevel(request, errorHandler.errorResult()(request)).map(Left(_))
-
+  private def makeAuthenticatedRequestWithRetrieval[A](
+    nino: String,
+    name: RetrievalName,
+    dateOfBirth: LocalDate,
+    request: MessagesRequest[A]
+  ): Future[Either[Result, AuthenticatedRequest[A]]] =
+    name match {
+      case RetrievalName(Some(forename), Some(surname)) =>
```

Comment:
this is where the auth team said that the full name goes in the `name` field. So we'd have to do something like:
```scala
    case RetrievalName(Some(fullName), _) =>
      fullName.split(' ').toList match {
        case firstName :: (_ :+ lastName) => // stuff
        case _                            => // error
      }
```

Reply:
not a great case class to work with....

Reply:
I also don't like this:

```scala
 private def makeAuthenticatedRequestWithItmpRetrieval[A](
    nino: String,
    name: ItmpName,
    dateOfBirth: LocalDate,
    request: MessagesRequest[A]
  ): Future[Either[Result, AuthenticatedRequest[A]]] =
    name match {
      case ItmpName(Some(givenName), _, Some(familyName)) =>
        Right(AuthenticatedRequest(NINO(nino), Name(givenName, familyName), DateOfBirth(dateOfBirth), request))
      case other =>
        logger.warn(s"Failed to retrieve the complete name for this user: $other")
        Left(errorHandler.errorResult()(request))
    }

  private def makeAuthenticatedRequestWithRetrieval[A](
    nino: String,
    name: RetrievalName,
    dateOfBirth: LocalDate,
    request: MessagesRequest[A]
  ): Future[Either[Result, AuthenticatedRequest[A]]] =
    name match {
      case RetrievalName(Some(forename), Some(surname)) =>
        Right(AuthenticatedRequest(NINO(nino), Name(forename, surname), DateOfBirth(dateOfBirth), request))
      case other =>
        logger.warn(s"Failed to retrieve the complete name for this user: $other")
        Left(errorHandler.errorResult()(request))
    }
```

I want to add a type parameter the method so that a sing;e method does this but the field names are different for each type.

Reply:
> this is where the auth team said that the full name goes in the `name` field. So we'd have to do something like:
> 
> ```scala
>     case RetrievalName(Some(fullName), _) =>
>       fullName.split(' ').toList match {
>         case firstName :: (_ :+ lastName) => // stuff
>         case _                            => // error
>       }
> ```

Yeah I realised I hadn't done that bit as soon as I popped out for lunch!

Reply:
the other option is to have a type signature which looks something like this:
```
def makeAuthenticatedRequestWithRetrieval[A](
    nino: String,
    firstName: String,
    lastName: String,
    dateOfBirth: LocalDate,
    request: MessagesRequest[A]
  ): Future[Either[Result, AuthenticatedRequest[A]]] =
```
and move the getting of the first and last name outside of the function.

Reply:
refactored this completely